lazy-pages/src/signal.rs

```diff
@@ -0,0 +1,255 @@
+        user_signal_handler(info)
+    }
+}
+
+#[derive(Debug)]
+pub(crate) struct ExceptionInfo {
+    /// Address where fault is occurred
+    pub fault_addr: *const (),
+    pub is_write: Option<bool>,
+}
+
+/// Before contract execution some pages from wasm memory buffer have been protected.
+/// When wasm executer tries to access one of these pages,
+/// OS emits sigsegv or sigbus or EXCEPTION_ACCESS_VIOLATION.
+/// Using OS signal info, this function identifies memory page, which is accessed,
+/// and process access for this page. See more in `process::process_lazy_pages`.
+/// After processing is done, OS returns execution to the same machine
+/// instruction, which cause signal. Now memory which this instruction accesses
+/// is not protected and with correct data.
+unsafe fn user_signal_handler_internal(
+    mut ctx: RefMut<LazyPagesExecutionContext>,
+    info: ExceptionInfo,
+) -> Result<(), Error> {
+    let native_addr = info.fault_addr as usize;
+    let is_write = info.is_write.ok_or(Error::ReadOrWriteIsUnknown)?;
+    let wasm_mem_addr = ctx.wasm_mem_addr.ok_or(Error::WasmMemAddrIsNotSet)?;
+
+    if native_addr < wasm_mem_addr {
+        return Err(Error::OutOfWasmMemoryAccess);
+    }
+
+    let offset =
+        u32::try_from(native_addr - wasm_mem_addr).map_err(|_| Error::OutOfWasmMemoryAccess)?;
+    let lazy_page = LazyPage::from_offset(offset);
+
+    let pages = if is_write {
+        utils::handle_psg_case_one_page(&mut ctx, lazy_page)?
+    } else {
+        lazy_page.iter_once()
+    };
+
+    let gas_left = if let Some(globals_config) = ctx.globals_config.as_ref() {
+        let gas = globals::apply_for_global(globals_config, GearGlobal::GasLimit, |_| Ok(None))?;
+        let allowance =
+            globals::apply_for_global(globals_config, GearGlobal::AllowanceLimit, |_| Ok(None))?;
+        Some(GasLeft { gas, allowance })
+    } else {
+        None
+    };
+
+    let handler = SignalAccessHandler { is_write, gas_left };
+    process::process_lazy_pages(ctx, handler, pages)
+}
+
+/// User signal handler. Logic can depends on lazy-pages version.
+/// See also "user_signal_handler_internal".
+pub(crate) unsafe fn user_signal_handler(info: ExceptionInfo) -> Result<(), Error> {
+    log::debug!("Interrupted, exception info = {:?}", info);
+    LAZY_PAGES_PROGRAM_CONTEXT.with(|ctx| user_signal_handler_internal(ctx.borrow_mut(), info))
+}
+
+struct SignalAccessHandler {
+    is_write: bool,
+    gas_left: Option<GasLeft>,
+}
+
+impl SignalAccessHandler {
+    fn sub_gas(gas_left: &mut GasLeft, amount: u64) -> Status {
+        log::trace!("charge amount = {amount}");
+        gas_left.gas = if let Some(gas) = gas_left.gas.checked_sub(amount) {
+            gas
+        } else {
+            return Status::GasLimitExceeded;
+        };
+        gas_left.allowance = if let Some(gas) = gas_left.allowance.checked_sub(amount) {
+            gas
+        } else {
+            return Status::GasAllowanceExceeded;
+        };
```

Comment:
looks like `let ... else` can be used there too

Reply:
Will discuss in above